ignite/metrics/mean_average_precision.py

```diff
@@ -0,0 +1,409 @@
+    """
+
+    def __init__(
+        self,
+        object_area_ranges: Optional[Dict[str, Tuple[float, float]]] = None,
+        num_detection_max: int = 100,
+        iou_thresholds: Optional[List[float]] = None,
+        rec_thresholds: Optional[List[float]] = None,
+        output_transform: Callable = lambda x: x,
+        device: Union[str, torch.device] = torch.device("cpu"),
+    ) -> None:
+
+        if object_area_ranges is None:
+            self.object_area_ranges = {
+                "all": (0.0, float("inf")),
+                "small": (0.0, 1024.0),
+                "medium": (1024.0, 9216.0),
+                "large": (9216.0, float("inf")),
+            }
+        else:
+            self.object_area_ranges = object_area_ranges
+            if "all" not in self.object_area_ranges:
+                self.object_area_ranges["all"] = (0, float("inf"))
+
+        self._check_object_area_ranges()
+
+        if num_detection_max < 1:
+            raise ValueError(f"Argument num_detection_max should be a positive integer, got {num_detection_max}")
+
+        self.num_detection_max = num_detection_max
+
+        if iou_thresholds is None:
+            iou_thresholds = [0.5, 0.55, 0.6, 0.65, 0.7, 0.75, 0.8, 0.85, 0.9, 0.95]
+        self.iou_thresholds = torch.tensor(iou_thresholds, device=device)
+
+        if rec_thresholds is None:
+            rec_thresholds = [i / 100 for i in range(101)]
+        self.rec_thresholds = torch.tensor(rec_thresholds, device=device)
+
+        class labels:
+            i_id = 0
+            i_category_id = 1
+            i_xmin = 2
+            i_area = 6
+            i_crowd = 7
+            g_ignore = 8
+            d_confidence = 8
+
+        self.labels = labels
+
+        super(MeanAveragePrecision, self).__init__(output_transform=output_transform, device=device)
+
+    def _check_object_area_ranges(self) -> None:
+        for area, area_range in self.object_area_ranges.items():
+            if len(area_range) != 2 or area_range[0] >= area_range[1]:
+                raise ValueError(
+                    f"object_area_ranges must be a dict associating to each key (str) a tuple of float values (a, b) "
+                    f"where a < b (got: key={area}, value={area_range})"
+                )
+
+    @reinit__is_reduced
+    def update(self, output: Tuple[torch.Tensor, torch.Tensor]) -> None:
+        y_pred_img, y_img = output
+
+        if len(y_pred_img.shape) != 2 or y_pred_img.shape[1] != 9:
+            raise ValueError(f"detections_tensor should be of size [num_detections, 9], got {y_pred_img.shape}")
+
+        if len(y_img.shape) != 2 or y_img.shape[1] != 9:
+            raise ValueError(f"ground_truths should be of size [num_detections, 9], got {y_img.shape}")
+
+        y_pred_img = y_pred_img.to(self._device)
+        y_img = y_img.to(self._device)
+
+        categories = set()
+        categorywise_y: defaultdict = defaultdict(lambda: torch.zeros(0, 9).to(self._device))
```

Comment:
Could you improve the design to avoid the stack ?

Reply:
I kept it as it is since it will get complicated....., did you check it again like we discussed on discord?